### PR TITLE
Use vector load/store in the normalization CUDA kernel

### PR DESCRIPTION
**Pull request by kaixih**
This PR is an attempt to use vector load/store for FP16 to optimize memory bandwidth utilization on supported GPUs. The target CUDA kernel is normalization which is used in softmax ops.

fyi @nluehr

### CODE REVIEW

**Comment by Ironicalink** [comment]
##I received this by email
@##Jobs:
build (3.5) cancelled (2 annotations)
build (3.6) cancelled (2 annotations)
build (3.7) failed (2 annotations)
build (3.8) failed (1 annotation)
—

##Hope this helps

**Comment by Ironicalink** [comment]
>##@Jules Vergogne, I don’t know what your trying to achieve?
##to : ##>:<><michael.ma.lortie@protonmail.com>
##@https://github.com/tensorflow/tensorflow/pull/38094#issuecomment-607385400

**Comment by Ironicalink** [comment]
> ##I received this by email
> 
> @##Jobs:
> 
> build (3.5) cancelled (2 annotations)
> 
> build (3.6) cancelled (2 annotations)
> 
> build (3.7) failed (2 annotations)
> 
> build (3.8) failed (1 annotation)
> 
> —
> 
> \<[git-pull-request]>/
> \:/
> ##Hope this helps
>/:\
> ##/$¥€&@&€¥$\

File: tensorflow/core/kernels/softmax_op_gpu.cu.cc [hunk 1]
```diff
@@ -72,28 +72,94 @@ struct softmax_traits<Eigen::half> {
   using accumulator_type = float;
 };
 
-template <typename T, typename U>
+template <typename T, typename U, int kUnroll>
```
Comment (rmlarsen) [inline tensorflow/core/kernels/softmax_op_gpu.cu.cc:75]: Nit: Do you expect to need other unroll values than 4 and 8? If no, we could replace the kUnroll parameter with a boolean "aligned" and just specialize on T = Eigen::half and aligned=true. It would save having two separate calls in the launcher below. I don't insist.

File: tensorflow/core/kernels/softmax_op_gpu.cu.cc [hunk 2]
```diff
@@ -72,28 +72,94 @@ struct softmax_traits<Eigen::half> {
...
+  float result[kUnroll];
+
+  if (tid * kUnroll + kUnroll - 1 < num_rows * num_cols) {
+    ulonglong2 logits_d =
+        *reinterpret_cast<const ulonglong2*>(logits + tid * kUnroll);
+    Eigen::half* logits_h = reinterpret_cast<Eigen::half*>(&logits_d);
+    ulonglong2 output_d;
+    Eigen::half* output_h = reinterpret_cast<Eigen::half*>(&output_d);
+
+    for (int i = 0; i < kUnroll; i++) {
```
Comment (rmlarsen) [inline tensorflow/core/kernels/softmax_op_gpu.cu.cc:134]: Would it be better to make logits_d a local variable and explicitly load kUnroll logits via logits_d instead of accessing logits_h[i] in the loop (like you do for the writes)? Or does  the compiler figure out how to coalesce those reads automatically?

Comment (houtoms) [reply tensorflow/core/kernels/softmax_op_gpu.cu.cc:134]: Thanks for your comments.
I think we already load the kUnroll logits via logits_d (line 129). What inside the loop (line 134) is accessing each half value (via logits_h[i]) to do the actual computation.

**Review by rmlarsen — APPROVED** [review]
